src-java/blue-green/src/main/java/org/openkilda/bluegreen/ZkWatchDog.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.data.Stat;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+@Slf4j
+public class ZkWatchDog extends ZkClient implements WatchDog, DataCallback {
+
+    private static final String SIGNAL = "signal";
+    @VisibleForTesting
+    protected String signalPath;
+    private Signal signal;
+
+    @VisibleForTesting
+    Set<LifeCycleObserver> observers = new HashSet<>();
+
+    @Builder
+    public ZkWatchDog(String id, String serviceName, String connectionString,
+                      int sessionTimeout, Signal signal) throws IOException {
+        super(id, serviceName, connectionString, sessionTimeout);
+
+        signalPath = getPaths(serviceName, id, SIGNAL);
+        if (signal == null) {
+            signal = Signal.NONE;
+        }
+        this.signal = signal;
+        initWatch();
+    }
+
+    @Override
+    void validateNodes() throws KeeperException, InterruptedException {
+        super.validateNodes();
+        ensureZNode(serviceName, id, SIGNAL);
+    }
+
+
+    @VisibleForTesting
+    void checkSignal() throws KeeperException, InterruptedException {
+        zookeeper.getData(signalPath, this, this, null);
+    }
+
+    @Override
+    void initWatch() {
+        try {
+            validateNodes();
+            checkSignal();
+        } catch (KeeperException | InterruptedException e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+
+    @Override
+    public void subscribe(LifeCycleObserver observer) {
+        observers.add(observer);
+    }
+
+    @Override
+    public void unsubscribe(LifeCycleObserver observer) {
+        if (observers.contains(observer)) {
+            observers.remove(observer);
+        }
+    }
+
+    @Override
+    public void process(WatchedEvent event) {
+        log.info("Received event: {}", event);
+        try {
+            if (!refreshConnection(event.getState()) && signalPath.equals(event.getPath())) {
+                checkSignal();
+            }
+        } catch (IOException e) {
+            log.error("Failed to read zk event: {}", e.getMessage(), e);
+        } catch (InterruptedException e) {
+            log.error("Zk event interrupted: {}", e.getMessage(), e);
+        } catch (KeeperException e) {
+            log.error("Zk keeper error: {}", e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void processResult(int rc, String path, Object ctx, byte[] data, Stat stat) {
+        log.debug("Received result on path: {}", path);
+        if (signalPath.equals(path) && data != null && data.length > 0) {
+            String signalString = new String(data);
+            try {
+                signal = Signal.valueOf(signalString);
+                notifyObservers();
+            } catch (Exception e) {
+                log.error("Received unknown signal: {}", signalString, e);
```

Comment:
If you will get any exception in `notifyObservers();` you will log (for example)
`Received unknown signal: START"`
But START is a valid signal

Reply:
changed